File: server/api/internal/infrastructure/websocket/dto.go
package websocket

type documentResponse struct {
	ID        string `json:"id"`
	Timestamp string `json:"timestamp"`
	Updates   []byte `json:"updates"`
	Version   uint64 `json:"version"`
}

type historyResponse struct {
	Timestamp string `json:"timestamp"`
	Updates   []byte `json:"updates"`
	Version   uint64 `json:"version"`
}

type rollbackRequest struct {
	DocID   string `json:"doc_id"`
	Version uint64 `json:"version"`
}

type createSnapshotRequest struct {
	DocID   string `json:"doc_id"`
	Name    string `json:"name"`
	Version uint64 `json:"version"`
}

type snapshotResponse struct {
	ID        string `json:"id"`
	Timestamp string `json:"timestamp"`
	Name      string `json:"name"`
	Updates   []byte `json:"updates"`
	Version   uint64 `json:"version"`
}

type importDocumentRequest struct {
	Data []int `json:"data"`
}

// snapshotItemResponse is one labelled snapshot; POST .../snapshots populates
// only ID and Label.
type snapshotItemResponse struct {
	Label     string `json:"label"`
	Timestamp string `json:"timestamp"`
	ID        int64  `json:"id"`
	Size      int64  `json:"size"`
}

type saveSnapshotRequest struct {
	Label string `json:"label"`
}

// snapshotStateResponse mirrors websocket-go's SnapshotStateResponse. Updates
// arrives as a JSON int array (the server's UpdateBytes marshaller); encoding/json
// decodes that into []byte as readily as it does base64.
type snapshotStateResponse struct {
	ID         string `json:"id"`
	Updates    []byte `json:"updates"`
	SnapshotID int64  `json:"snapshot_id"`
}

File: server/websocket/src/presentation/http/dto.rs
use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize)]
pub struct DocumentResponse {
    pub id: String,
    pub updates: Vec<u8>,
    pub version: u64,
    pub timestamp: String,
}

#[derive(Serialize, Deserialize)]
pub struct GetLatestRequest {
    pub doc_id: String,
}

#[derive(Serialize, Deserialize)]
pub struct GetHistoryRequest {
    pub doc_id: String,
}

#[derive(Serialize, Deserialize)]
pub struct RollbackRequest {
    pub doc_id: String,
    pub version: u64,
}

#[derive(Serialize, Deserialize)]
pub struct CreateSnapshotRequest {
    pub doc_id: String,
    pub version: u64,
    pub name: Option<String>,
}

#[derive(Serialize, Deserialize)]
pub struct SnapshotResponse {
    pub id: String,
    pub updates: Vec<u8>,
    pub version: u64,
    pub timestamp: String,
    pub name: Option<String>,
}

#[derive(Serialize, Deserialize)]
pub struct HistoryResponse {
    pub updates: Vec<u8>,
    pub version: u64,
    pub timestamp: String,
}

#[derive(Serialize)]
pub struct HistoryMetadataResponse {
    pub version: u64,
    pub timestamp: String,
}

#[derive(Serialize, Deserialize)]
pub struct ImportDocumentRequest {
    pub data: Vec<u8>,
}

#[derive(Serialize)]
pub struct CleanupResponse {
    pub deleted: usize,
}

#[derive(Serialize)]
pub struct BatchCleanupResponse {
    pub docs_processed: usize,
    pub total_deleted: usize,
}
